client/web/src/codeintel/location.ts

```diff
@@ -15,27 +15,182 @@ export interface Location {
+        this.locations = [locations[0]]
+        for (const [i, loc] of locations.entries()) {
+            if (i === 0) {
+                continue
+            }
+            this.tryAdd(loc)
+        }
+    }
+
+    /**
+     * Attempt to add the Location to this group without mixing
+     * search-based and precise Locations.
+     *
+     * If one attempts to mix them, precise locations will be preferred.
+     */
+    private tryAdd(location: Location): void {
+        if (this.precise && !location.precise) {
+            return
+        }
+        if (!this.precise && location.precise) {
+            this.precise = true
+            this.locations = [location]
+            return
+        }
+        if (location.file !== this.path) {
+            throw new Error('pre-condition failure')
+        }
+        if (location.precise !== this.precise) {
+            throw new Error('already handled precise same-ness check earlier')
+        }
+        this.locations.push(location)
+    }
+
+    /** Do not modify the returned Array. */
+    public getLocations(): Location[] {
+        return this.locations
+    }
+
+    public get quality(): LocationGroupQuality {
+        return this.precise ? 'PRECISE' : 'SEARCH-BASED'
+    }
 }
 
-export type LocationGroupQuality = 'PRECISE' | 'SEARCH-BASED'
+/**
+ * Type to store locations grouped by (repo, file) pairs.
+ *
+ * This type is specialized for use in the reference panel code.
+ * So if a given (repo, file) pair contains both search-based Locations
+ * and precise Locations, the search-based Locations are discarded.
+ */
+export class LocationsGroup {
+    /** Invariant: `locationsCount` is the sum of sizes of Location arrays in `groups` */
+    private locationsCount: number
+    /** Invariant: Every Location stored in the group has a distinct URL. */
+    private groups: LocationsGroupedByRepo[]
+
+    constructor(locations: Location[]) {
+        this.locationsCount = 0
+        this.groups = []
 
-export const locationGroupQuality = (group: LocationGroup): LocationGroupQuality => {
-    if (group.locations.length === 0) {
-        throw new Error(`No locations in ${group.path}`)
+        const urlsSeen = new Set<string>()
+        const repoMap = new Map<string, Map<string, Location[]>>()
+        for (const loc of locations) {
+            if (urlsSeen.has(loc.url)) {
+                continue
+            }
+            urlsSeen.add(loc.url)
+            const pathToLocMap = repoMap.get(loc.repo)
+            if (pathToLocMap) {
+                const fileLocs = pathToLocMap.get(loc.file)
+                if (fileLocs) {
+                    fileLocs.push(loc)
+                } else {
+                    pathToLocMap.set(loc.file, [loc])
+                }
+            } else {
+                const pathToLocMap = new Map<string, Location[]>()
+                pathToLocMap.set(loc.file, [loc])
+                repoMap.set(loc.repo, pathToLocMap)
+            }
```

Comment:
I find the following easier to understand (don't know if this "pattern" has a name):

```suggestion
            let pathToLocMap = repoMap.get(loc.repo)
            if (!pathToLocMap) {
                pathToLocMap = new Map<string, Location[]>()
            }
            const fileLocs = pathToLocMap.get(loc.file)
            if (fileLocs) {
              fileLocs.push(loc)
            } else {
              pathToLocMap.set(loc.file, [loc])
            }
```

Reply:
Simplified the code as requested.